### PR TITLE
Slack integration poc

### PR DESCRIPTION
**Pull request by Neptunao**

### CODE REVIEW

File: src/KawaBot.Server/OAuthController.fs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+open System.Net
+open Newtonsoft.Json
+open Environment
+open System.Collections.Generic
+open Newtonsoft.Json.Linq
+open System.Text
+open System.Threading.Tasks
+open System
+
+[<JsonObject>]
```
Comment (sirIrishman) [inline src/KawaBot.Server/OAuthController.fs]: Lets create `module OAuthController =` for this file. As it's done in the `HealthController`. What do you think?

File: src/KawaBot.Server/OAuthController.fs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+type SlackError =
+    {
+        [<field: JsonProperty("ok") >]
+        Ok: bool;
+        [<field: JsonProperty("error") >]
+        Error: string;
+    }
+
+[<JsonObject>]
+type OAuthBot = {
```
Comment (sirIrishman) [inline src/KawaBot.Server/OAuthController.fs]: Formatting of `{}` is inconsistent.

File: src/KawaBot.Server/OAuthController.fs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        UserId: string;
+        [<field: JsonProperty("team_name") >]
+        TeamName: string;
+        [<field: JsonProperty("team_id") >]
+        TeamId: string;
+        [<field: JsonProperty("bot") >]
+        Bot: OAuthBot;
+    }
+
+module OAuthHelper =
```
Comment (sirIrishman) [inline src/KawaBot.Server/OAuthController.fs]: Lets move it to a separate file.

File: src/KawaBot.Server/OAuthController.fs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    let GetOAuthURL(code: string): string =
+            let urlArgs = [
```
Comment (sirIrishman) [inline src/KawaBot.Server/OAuthController.fs]: Double indentation.

File: src/KawaBot.Server/OAuthController.fs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                "https://slack.com/api/oauth.access";
+                sprintf "?client_id=%s" Environment.ClientID;
+                sprintf  "&client_secret=%s" Environment.ClientSecret;
+                sprintf "&code=%s" code;
+            ] 
+            urlArgs |> String.concat ""
+
+type IHTTPRequester =
```
Comment (sirIrishman) [inline src/KawaBot.Server/OAuthController.fs]: Lets use the Pascal case for abbreviations longer than 3 chars: `HTTP` -> `Http`.

File: src/KawaBot.Server/OAuthController.fs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+            let urlArgs = [
+                "https://slack.com/api/oauth.access";
+                sprintf "?client_id=%s" Environment.ClientID;
+                sprintf  "&client_secret=%s" Environment.ClientSecret;
+                sprintf "&code=%s" code;
+            ] 
+            urlArgs |> String.concat ""
+
+type IHTTPRequester = 
+    abstract member HTTPGet: url: string -> HttpResponseMessage
```
Comment (sirIrishman) [inline src/KawaBot.Server/OAuthController.fs]: Lets use the Pascal case for abbreviations longer than 3 chars: `HTTP` -> `Http`.

File: src/KawaBot.Server/OAuthController.fs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+type HTTPRequester() =
```
Comment (sirIrishman) [inline src/KawaBot.Server/OAuthController.fs]: Lets use the Pascal case for abbreviations longer than 3 chars: `HTTP` -> `Http`.

File: src/KawaBot.Server/OAuthController.fs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    interface IHTTPRequester with
```
Comment (sirIrishman) [inline src/KawaBot.Server/OAuthController.fs]: Lets use the Pascal case for abbreviations longer than 3 chars: `HTTP` -> `Http`.

File: src/KawaBot.Server/OAuthController.fs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        member this.HTTPGet (url: string): HttpResponseMessage =
```
Comment (sirIrishman) [inline src/KawaBot.Server/OAuthController.fs]: Lets use the Pascal case for abbreviations longer than 3 chars: `HTTP` -> `Http`.

File: src/KawaBot.Server/OAuthController.fs [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+            use client = new HttpClient()
+            Logger.Debug("performing request GET %s\n", url)
+            client.GetAsync(url).Result
+
+
+type OAuthController(requester: IHTTPRequester) =
```
Comment (sirIrishman) [inline src/KawaBot.Server/OAuthController.fs]: Lets use the Pascal case for abbreviations longer than 3 chars: `HTTP` -> `Http`.

File: src/KawaBot.Server/OAuthController.fs [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+                            let err = JsonConvert.DeserializeObject<SlackError>(body)
+                            Logger.Error(err.Error)
+                        | true -> 
+                            let oauthResp = JsonConvert.DeserializeObject<OAuthSuccessResponse>(body)
+                            assert (oauthResp.Bot.BotAccessToken <> "")
+                            Environment.BotToken <- oauthResp.Bot.BotAccessToken
+                            Logger.Info("successfully obtain the bot token\n")
+                | false, _ -> 
+                    Logger.Error("Invalid OAuth format")
+                    resp.StatusCode <- 400
```
Comment (sirIrishman) [inline src/KawaBot.Server/OAuthController.fs]: Please, use the `HttpStatusCode` enum.

File: src/KawaBot.Feeder/Feeder.fs [hunk 12]
```diff
@@ -2,6 +2,8 @@
 
 open System
 open System.Net.Http
+open Newtonsoft.Json
+open System.Text
```
Comment (sirIrishman) [inline src/KawaBot.Feeder/Feeder.fs]: Please sort usings.

File: src/KawaBot.Server/OAuthController.fs [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+open Http
+open System.Net.Http
+open System.Net
+open Newtonsoft.Json
+open Environment
+open System.Collections.Generic
+open Newtonsoft.Json.Linq
+open System.Text
+open System.Threading.Tasks
+open System
```
Comment (sirIrishman) [inline src/KawaBot.Server/OAuthController.fs]: Please sort usings.

File: src/KawaBot.Server/OAuthController.fs [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    abstract member HTTPGet: url: string -> HttpResponseMessage
+
+type HTTPRequester() =
+    interface IHTTPRequester with
+        member this.HTTPGet (url: string): HttpResponseMessage =
+            use client = new HttpClient()
+            Logger.Debug("performing request GET %s\n", url)
+            client.GetAsync(url).Result
+
+
```
Comment (sirIrishman) [inline src/KawaBot.Server/OAuthController.fs]: Excessive new lines.

File: src/KawaBot.Server/OAuthController.fs [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+type OAuthController(requester: IHTTPRequester) =
+    let _requester = requester
+
+    member this.GetUrlHandler() =
+        new UrlHandler("oauth", fun req resp -> async {
+            let url = OAuthHelper.GetOAuthURL(req.QueryString.Get("code"))
+            use authResp = _requester.HTTPGet(url)
+            Logger.Debug("success, status %s\n", (authResp.StatusCode.ToString()))
+            let! body = authResp.Content.ReadAsStringAsync() |> Async.AwaitTask
+            let jobj = JObject.Parse(body)
```
Comment (sirIrishman) [inline src/KawaBot.Server/OAuthController.fs]: Rename it please to something meaningful.

File: src/KawaBot.Tests/BasicControllerTests.fs [hunk 16]
```diff
@@ -4,20 +4,36 @@ open System.Net
 open KawaBot.Feeder
 open KawaBot.Server
 open NUnit.Framework
+open Moq
+open System.Net.Http
+open System.Text
+open Newtonsoft.Json
```
Comment (sirIrishman) [inline src/KawaBot.Tests/BasicControllerTests.fs]: Please sort usings.

File: src/KawaBot.Server/SlackVerificationController.fs [hunk 17]
```diff
@@ -0,0 +1,37 @@
+﻿namespace KawaBot.Server
+
+open System.IO
+open Http
+open Newtonsoft.Json
+open System.Collections.Generic
+open Newtonsoft.Json.Linq
```
Comment (sirIrishman) [inline src/KawaBot.Server/SlackVerificationController.fs]: Please sort usings.

File: src/KawaBot.Tests/BasicControllerTests.fs [hunk 18]
```diff
@@ -4,20 +4,36 @@ open System.Net
...
 type BasicControllerTests =
     val private _server: Http.HttpListener
     val private _feeder: Feeder
+    val private _httpRequsterMock: Mock<IHTTPRequester>
 
     new() as this =
         {
             _server = new Http.HttpListener("http://*:80/")
             _feeder = new Feeder(80)
+            _httpRequsterMock = Mock<IHTTPRequester>()
```
Comment (sirIrishman) [inline src/KawaBot.Tests/BasicControllerTests.fs]: Lets use the Pascal case for abbreviations longer than 3 chars: `HTTP` -> `Http`.

File: src/KawaBot.Tests/BasicControllerTests.fs [hunk 19]
```diff
@@ -4,20 +4,36 @@ open System.Net
...
+open System.Net.Http
+open System.Text
+open Newtonsoft.Json
 
 [<TestFixture>]
 [<Category(Categories.Integration)>]
 type BasicControllerTests =
     val private _server: Http.HttpListener
     val private _feeder: Feeder
+    val private _httpRequsterMock: Mock<IHTTPRequester>
```
Comment (sirIrishman) [inline src/KawaBot.Tests/BasicControllerTests.fs]: Lets use the Pascal case for abbreviations longer than 3 chars: `HTTP` -> `Http`.

File: src/KawaBot.Tests/BasicControllerTests.fs [hunk 20]
```diff
@@ -4,20 +4,36 @@ open System.Net
...
             |> this._server.Register
+            SlackVerificationController.GetUrlHandler()
+            |> this._server.Register
+            let oauth = new OAuthController(this._httpRequsterMock.Object)
+            oauth.GetUrlHandler() |> this._server.Register
+
+    member this.StubOAuthController(body: string) =
+        let resp = new HttpResponseMessage()
+        resp.Content <- new ByteArrayContent(Encoding.UTF8.GetBytes(body))
+        this._httpRequsterMock.Setup(fun req -> req.HTTPGet(It.IsAny<string>()))
```
Comment (sirIrishman) [inline src/KawaBot.Tests/BasicControllerTests.fs]: Lets use the Pascal case for abbreviations longer than 3 chars: `HTTP` -> `Http`.

File: src/KawaBot.Tests/BasicControllerTests.fs [hunk 21]
```diff
@@ -37,3 +53,45 @@ type BasicControllerTests =
...
         Assert.That(response.Content.ReadAsStringAsync().Result, Is.EqualTo("OK"))
+
+    [<Test>]
+    member this.TestVerification() = 
+        let v = {
+            Token = "";
+            Challenge = "challenge";
+            Type = "";
+        }
+        let response = this._feeder.PostJSON("/slack", v) |> Async.RunSynchronously
```
Comment (sirIrishman) [inline src/KawaBot.Tests/BasicControllerTests.fs]: Lets use the Pascal case for abbreviations longer than 3 chars: `JSON` -> `Json`.

File: src/KawaBot.Feeder/Feeder.fs [hunk 22]
```diff
@@ -19,3 +21,13 @@ type Feeder =
         new HttpRequestMessage(defaultArg method HttpMethod.Get, message)
         |> this._client.SendAsync
         |> Async.AwaitTask
+
+    member this.PostJSON(message: string, body: 'a) =
```
Comment (sirIrishman) [inline src/KawaBot.Feeder/Feeder.fs]: Lets use the Pascal case for abbreviations longer than 3 chars: `JSON` -> `Json`.

File: src/KawaBot.Feeder/Feeder.fs [hunk 23]
```diff
@@ -19,3 +21,13 @@ type Feeder =
...
+        let msg = new HttpRequestMessage(HttpMethod.Post, message)
+        let bodyJson = JsonConvert.SerializeObject(body, Formatting.Indented)
+        let bac = new ByteArrayContent(Encoding.UTF8.GetBytes(bodyJson))
```
Comment (sirIrishman) [inline src/KawaBot.Feeder/Feeder.fs]: Rename it please to something meaningful.

File: src/KawaBot.Server/SlackVerificationController.fs [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+            let responseStr = inputReader.ReadToEnd()
+            let jobj = JObject.Parse(responseStr)
+            match jobj.TryGetValue("challenge") with
+                | true, challenge -> 
+                    //TODO check (verification) token field
+                    let verification = JsonConvert.DeserializeObject<SlackVerification>(responseStr)
+                    printfn "Got verification, challenge='%s'" verification.Challenge
+                    writer.Write(verification.Challenge)
+                | false, _ -> 
+                   printfn "got event: %s" responseStr
```
Comment (coffeecup-winner) [inline src/KawaBot.Server/SlackVerificationController.fs]: the indentation should be 1 space more

File: src/KawaBot.Server/SlackVerificationController.fs [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+module SlackVerificationController =
+    let GetUrlHandler() =
+        new UrlHandler("slack", fun req resp -> async {
+            assert (req.HttpMethod = "POST")
+            assert (req.ContentType = "application/json")
+            use inputReader = new StreamReader(req.InputStream)
+            use writer = new StreamWriter(resp.OutputStream)
+            let responseStr = inputReader.ReadToEnd()
```
Comment (coffeecup-winner) [inline src/KawaBot.Server/SlackVerificationController.fs]: OPT: rename responseStr -> body to be consistent with OAuth

File: src/KawaBot.Tests/BasicControllerTests.fs [hunk 26]
```diff
@@ -4,20 +4,36 @@ open System.Net
 open KawaBot.Feeder
 open KawaBot.Server
 open NUnit.Framework
+open Moq
```
Comment (coffeecup-winner) [inline src/KawaBot.Tests/BasicControllerTests.fs]: sort opens

File: src/KawaBot.Tests/BasicControllerTests.fs (hunk 19 above)
Comment (coffeecup-winner) [inline src/KawaBot.Tests/BasicControllerTests.fs]: typo

File: src/KawaBot.Tests/BasicControllerTests.fs [hunk 27]
```diff
@@ -37,3 +53,45 @@ type BasicControllerTests =
         let response = this._feeder.Send("/health") |> Async.RunSynchronously
         Assert.That(response.StatusCode, Is.EqualTo(HttpStatusCode.OK))
         Assert.That(response.Content.ReadAsStringAsync().Result, Is.EqualTo("OK"))
+
+    [<Test>]
```
Comment (coffeecup-winner) [inline src/KawaBot.Tests/BasicControllerTests.fs]: move this test to SlackControllerTests.fs

File: src/KawaBot.Tests/BasicControllerTests.fs [hunk 28]
```diff
@@ -37,3 +53,45 @@ type BasicControllerTests =
...
+        let v = {
+            Token = "";
+            Challenge = "challenge";
+            Type = "";
+        }
+        let response = this._feeder.PostJSON("/slack", v) |> Async.RunSynchronously
+        Assert.That(response.StatusCode, Is.EqualTo(HttpStatusCode.OK))
+        Assert.That(response.Content.ReadAsStringAsync().Result, Is.EqualTo(v.Challenge))
+
+    [<Test>]
```
Comment (coffeecup-winner) [inline src/KawaBot.Tests/BasicControllerTests.fs]: move these tests to OAuthControllerTests.fs

File: src/KawaBot.Tests/BasicControllerTests.fs [hunk 29]
```diff
@@ -37,3 +53,45 @@ type BasicControllerTests =
         let response = this._feeder.Send("/health") |> Async.RunSynchronously
         Assert.That(response.StatusCode, Is.EqualTo(HttpStatusCode.OK))
         Assert.That(response.Content.ReadAsStringAsync().Result, Is.EqualTo("OK"))
+
+    [<Test>]
+    member this.TestVerification() = 
+        let v = {
+            Token = "";
```
Comment (coffeecup-winner) [inline src/KawaBot.Tests/BasicControllerTests.fs]: no ;s

File: src/KawaBot.Tests/BasicControllerTests.fs [hunk 30]
```diff
@@ -37,3 +53,45 @@ type BasicControllerTests =
...
+    [<Test>]
+    member this.TestOAuthInvalidResponse() =
+        this.StubOAuthController("{}") |> ignore
+        let response = this._feeder.Send("/oauth?code=123") |> Async.RunSynchronously
+        Assert.That(response.StatusCode, Is.EqualTo(HttpStatusCode.BadRequest))
+
+    [<Test>]
+    member this.TestOAuthGetsBotToken() =
+        let res = {
+            Ok = true;
```
Comment (coffeecup-winner) [inline src/KawaBot.Tests/BasicControllerTests.fs]: no ;s

File: src/KawaBot.Tests/BasicControllerTests.fs [hunk 31]
```diff
@@ -37,3 +53,45 @@ type BasicControllerTests =
...
+            TeamId = "";
+            Bot = {
+                BotUserId = "";
+                BotAccessToken = "VerySecretBotToken";
+            }
+        }
+        this.StubOAuthController(JsonConvert.SerializeObject(res, Formatting.Indented)) |> ignore
+        let response = this._feeder.Send("/oauth?code=123") |> Async.RunSynchronously
+        Assert.That(response.StatusCode, Is.EqualTo(HttpStatusCode.OK))
+        Assert.That(Environment.BotToken, Is.EqualTo(res.Bot.BotAccessToken))
```
Comment (coffeecup-winner) [inline src/KawaBot.Tests/BasicControllerTests.fs]: add newline

File: src/KawaBot.Tests/BasicControllerTests.fs [hunk 32]
```diff
@@ -37,3 +53,45 @@ type BasicControllerTests =
...
+            Scope = "";
+            UserId = "";
+            TeamName = "";
+            TeamId = "";
+            Bot = {
+                BotUserId = "";
+                BotAccessToken = "VerySecretBotToken";
+            }
+        }
+        this.StubOAuthController(JsonConvert.SerializeObject(res, Formatting.Indented)) |> ignore
```
Comment (coffeecup-winner) [inline src/KawaBot.Tests/BasicControllerTests.fs]: why do you need formatting there? 😃

File: .gitignore [hunk 33]
```diff
@@ -190,6 +190,7 @@ ClientBin/
 *.dbproj.schemaview
 *.pfx
 *.publishsettings
+*.orig
```
Comment (Neptunao) [inline .gitignore]: And why not? They can remain for some reason.

File: src/KawaBot.Server/OAuthController.fs [hunk 34]
```diff
@@ -0,0 +1,97 @@
...
+            match jobj.TryGetValue("ok") with
```
Comment (Neptunao) [inline src/KawaBot.Server/OAuthController.fs]: Ok, will try

File: src/KawaBot.Tests/BasicControllerTests.fs (hunk 32 above)
Comment (Neptunao) [inline src/KawaBot.Tests/BasicControllerTests.fs]: For debugging.

**Pull request closed without merging**